Migrate daemon to libgpiod

hangeza opened the pull request:
All (but one minor) functionality implemented and successfully tested so far based on libgpiod instead of pigpiod. The low-level interrupt handler now sends signals to an intermediate instance of RateBuffer, which provides protection against sustained high count rates. Following functionalities still need to be implemented or taken care of:
- user muonpi must be in gpio group (not yet fixed in postinstall script)
- automatic measurement of the system time against the timepulse not yet implemented as in the previous pigpiod-based version. However, this measurement (affecting only the TPTimeDiff histogram ) only makes sense, when the system is time-synchronized (with eg. chrony or ntpd) and was not regarded up to now by anyone.

closes #40

hangeza committed:
Merge pull request #41 from MuonPi/dev

Dev

hangeza committed:
Started modifying pigpiodhandler class to switch over to libgpiod. Not finished yet...code will not compile.

hangeza committed:
Cleaned up PigpiodHandler class a bit and implemented one signal to forward a detected event to the daemon. With this, all basic functions of the daemon seem to work.
Not working yet and to be implemented in the new gpio readout scheme:
- pulse rate limitation for high freq. bursts
- sampling trigger signals (for triggering the ADC readout)
- time and time difference measurements for the population of two histos

Note: For the daemon to work with the new libgpiod based readout, several things have to be changed from the pigpiod scheme:
- the user who is executing the daemon (usually muonuser) must be added to the system group 'gpio' with 'sudo usermod -a -G gpio muonuser'
- the pigpiod daemon must be switched off
- the libgpiod-dev package must be installed on the system prior to compilation

hangeza committed:
added start() and stop() slots to PigpiodHandler class to always have a defined run status of the internal thread

hangeza committed:
Implemented pure usage of libgpiod in PigpiodHandler class. Signals from the former callback function are not yet forwarded, but the core mechanism of event detection works. Pins may be registered/unregistered during runtime and set/get gpio functions also seem to work. Note, that the executing user must be member of gpio group.

hangeza committed:
small change in debug output of detected edges

hangeza committed:
Further implemented functionality equal to the old pigpiod implementation:
- implemented trigger for ADC sampling
- implemented GPIO inhibit
- default eventTrigger signal is now GPIO_UNDEFINED_PIN, i.e. no trigger set up. This still has to be implemented as valid selection in the GUI

hangeza committed:
Added RateBuffer class for handling of all GPIO rates. Still to be implemented for the tracking of the XOR and AND rates in the daemon.

hangeza committed:
Embedded RateBuffer in daemon. All rate handling should be supervised by an instance of this class now. RateBuffer contains a smart rate throttling which prevents lock-up of daemon and GUI at excessive rates.
This branch should now privide equal or better functionality compared to the pigpiod daemon.
However, the following features are not yet implemented:
- Measurement of the system time deviation from GNSS (timpulse) time
- Measurement of the GPIO event interval (as in the histogram under statistics tab of the GUI with the same name). This may be implemented easily with the help of the RateBuffer class.

hangeza committed:
Added measurement of gpio interval time (in histogram "gpioEventInterval"). Adjusted scaling of several histograms.

hangeza committed:
Renamed enum GPIO_PIN to the more reasonable GPIO_SIGNAL, since this signal is explicitly not bound to a fixed GPIO pin.
Corrected some wrong assignments in the GPIO pin list for HW Ver 1.

hangeza committed:
fixed wrong position of default argument in constructor of i2cdevice class

hangeza committed:
Merge branch 'dev' into migrate-daemon-to-libgpiod

hangeza committed:
added missing linkage against pthread in daemon.cmake

hangeza committed:
Updated preinst script to add muonuser to gpio group. This is required for running the daemon with access to the gpio kernel subsystem.

hangeza committed:
- modified all control scripts for prerequisites etc. wrt libgpiod
- added measurement of time pulse to system time

hangeza committed:
adjust measured TPDiff values symmetrically around 0, i.e. to fall in the range -0.5s...+0.5s, instead of 0...1s as before

dtreffenstaedt reviewed (changes requested):
Please don't comment out code but directly delete it instead.

dtreffenstaedt commented on source/daemon/include/daemon.h:
```diff
@@ -110,7 +112,7 @@ struct RateScanInfo {
 struct RateScan {
 //	void addScanPoint(double scanpar, double a_rate) { scanMap[scanpar].append(a_rate); }
```
```suggestion
```

dtreffenstaedt commented on source/daemon/include/pigpiodhandler.h:
```diff
@@ -20,44 +26,56 @@ class PigpiodHandler : public QObject
+	// can't make it private because of access of PigpiodHandler with global pointer
+	QDateTime startOfProgram, lastSamplingTime; // the exact time when the program starts (Utc)
+	QElapsedTimer elapsedEventTimer;
+//	GPIO_SIGNAL samplingTriggerSignal=EVT_XOR;
```
```suggestion
```

dtreffenstaedt commented on source/daemon/include/pigpiodhandler.h:
```diff
@@ -20,44 +26,56 @@ class PigpiodHandler : public QObject
-
-    // spi related signals
-    void spiData(uint8_t reg, std::string data);
+    //void samplingTrigger();
+	//void eventInterval(quint64 nsecs);
+	//void timePulseDiff(qint32 usecs);
```
```suggestion
```

dtreffenstaedt commented on source/daemon/include/pigpiodhandler.h:
```diff
@@ -20,44 +26,56 @@ class PigpiodHandler : public QObject
+	bool setPinBias(unsigned int gpio, std::uint8_t pin_bias);
+    bool setPinState(unsigned int gpio, bool state);
+    void setSamplingTriggerSignal(unsigned int gpio);
+//    void setSamplingTriggerSignal(GPIO_SIGNAL signalName) { samplingTriggerSignal=signalName; }
```
```suggestion
```

dtreffenstaedt commented on source/daemon/include/pigpiodhandler.h:
```diff
@@ -79,8 +97,26 @@ public slots:
 	bool inhibit=false;
 	int verbose=0;
+	gpiod_chip* fChip { nullptr };
+//	std::shared_ptr<struct gpiod_chip> fChip { nullptr };
```
```suggestion
```

dtreffenstaedt commented on source/daemon/include/pigpiodhandler.h:
```diff
@@ -79,8 +97,26 @@ public slots:
+	gpiod_chip* fChip { nullptr };
+//	std::shared_ptr<struct gpiod_chip> fChip { nullptr };
+	std::map<unsigned int, gpiod_line*> fInterruptLineMap { };
+//	std::map<unsigned int, std::shared_ptr<gpiod_line>> fInterruptLineMap { };
```
```suggestion
```

dtreffenstaedt commented on source/daemon/include/pigpiodhandler.h:
```diff
@@ -79,8 +97,26 @@ public slots:
+//	std::map<unsigned int, std::shared_ptr<gpiod_line>> fInterruptLineMap { };
+	gpiod_line_bulk fInterruptLineBulk;
+	std::map<unsigned int, gpiod_line*> fLineMap { };
+//	std::map<unsigned int, std::shared_ptr<gpiod_line>> fLineMap { };
```
```suggestion
```

dtreffenstaedt commented on source/daemon/include/pigpiodhandler.h:
```diff
@@ -79,8 +97,26 @@ public slots:
+	bool fThreadRunning { false };
+	std::unique_ptr<std::thread> fThread { nullptr };
+	std::mutex fMutex;
+/*	
+	gpiod::chip fChip;
+	gpiod::line_bulk fInterruptLineBulk;
+	std::map<unsigned int, gpiod::line> fLineMap;
+*/
 //	QVector<unsigned int> gpioPins;
```
```suggestion
```

dtreffenstaedt commented on source/daemon/src/ratebuffer.cpp:
```diff
@@ -0,0 +1,91 @@
+		auto last_event_time = buffermap[gpio].eventbuffer.back();
+		buffermap[gpio].last_interval = std::chrono::duration_cast<std::chrono::nanoseconds>( now - last_event_time );
+		if ( now - last_event_time < MAX_DEADTIME ) {
+//			std::cout << "now-last:"<<(now-last_event_time)/1us<<" dt="<<buffermap[gpio].current_deadtime.count()<<std::endl;
```
```suggestion
```

dtreffenstaedt commented on source/gui/src/mainwindow.cpp:
```diff
@@ -398,7 +398,7 @@ void MainWindow::receivedTcpMessage(TcpMessage tcpMessage) {
 //	if (msgID == gpioPinSig) {
```
```suggestion
```

dtreffenstaedt commented on source/gui/src/parametermonitorform.cpp:
```diff
@@ -189,7 +189,7 @@ void ParameterMonitorForm::onPreampSwitchReceived(uint8_t channel, bool state)
-void ParameterMonitorForm::onTriggerSelectionReceived(GPIO_PIN signal)
+void ParameterMonitorForm::onTriggerSelectionReceived(GPIO_SIGNAL signal)
 {
     //if (GPIO_PIN_NAMES.find(signal)==GPIO_PIN_NAMES.end()) return;
```
```suggestion
```

dtreffenstaedt commented on source/gui/src/status.cpp:
```diff
@@ -162,7 +162,7 @@ void Status::clearRatePlot()
-void Status::onTriggerSelectionReceived(GPIO_PIN signal)
+void Status::onTriggerSelectionReceived(GPIO_SIGNAL signal)
 {
 //    if (GPIO_PIN_NAMES.find(signal)==GPIO_PIN_NAMES.end()) return;
```
```suggestion
```

dtreffenstaedt commented on source/daemon/config/preinst:
```diff
@@ -1,6 +1,6 @@
 
 mkdir -p /var/muondetector
-useradd muonuser -g users -G dialout,pi,i2c,users -s /usr/sbin/nologin -r -N -M -b /var/muondetector || echo "User already exists"
+useradd muonuser -g users -G dialout,pi,i2c,users,gpio -s /usr/sbin/nologin -r -N -M -b /var/muondetector || adduser muonuser gpio
```
```suggestion
if id muonuser &> /dev/null; then
    # transitional code until all older versions of muondetector-daemon are updated.
    echo "muonuser exists, adding to group gpio."
    adduser muonuser gpio
else
    useradd muonuser -g users -G dialout,pi,i2c,users,gpio -s /usr/sbin/nologin -r -N -M -b /var/muondetector
 fi
```

dtreffenstaedt reviewed (changes requested):
Add a more sturdy preinst script to handle older installations where just the user needs to be added to a new group

hangeza committed:
Update source/daemon/include/daemon.h

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/daemon/include/pigpiodhandler.h

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/daemon/include/pigpiodhandler.h

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/daemon/include/pigpiodhandler.h

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/daemon/include/pigpiodhandler.h

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/daemon/include/pigpiodhandler.h

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/daemon/include/pigpiodhandler.h

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/daemon/include/pigpiodhandler.h

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/daemon/src/ratebuffer.cpp

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/gui/src/mainwindow.cpp

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/gui/src/parametermonitorform.cpp

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/gui/src/status.cpp

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Update source/daemon/config/preinst

Co-authored-by: Daniel Treffenstädt <52167116+dtreffenstaedt@users.noreply.github.com>

hangeza committed:
Some clean-up and minor mods:
- removed separate the samplingTrigger handling in PigpiodHandler
- merged slots for gpio events in daemon into one onGpioPinEvent() slot
- removed unregistering interrupts in PigpiodHandler before registering them again. This would be the cleanest procedure but somehow it doesn't work from gpiod side

hangeza committed:
- Renamed signal and throttledSignal signals in PigpioHandler and RateBuffer classes to more descriptive event and filteredEvent.
- Added transmission of interrupt timestamp to event and filteredEvent signals.

hangeza committed:
Further work on libgpiod functionality for event handling:
- moved all event handlers to separately spawned threads for each interrupt line
- added map to store the last time intervals between events from different gpios in RateBuffer class
- changed filling of gpioTimeIntervalShort histogram to time intervals measured between the currently selected eventTrigger and the rising edge of TIME_MEAS_OUT signal

hangeza committed:
Fixed merge conflicts and adapted code at some places for proper compile success

hangeza committed:
deleted some now obsolete files

hangeza committed:
Merge branch 'dev' into migrate-daemon-to-libgpiod

hangeza committed:
Fixed some lines which slipped through at the last merge. Compiles now

hangeza committed:
Restored muondetector-daemon service file which was corrupted from a prior merge conflict. Added ratebuffer.cpp file which was not commited properly during the last big PR.

hangeza committed:
renamed pin_bias to bias_flags

hangeza committed:
removed commented-out lines

hangeza committed:
added suffix string "libgpiod" to version strings (e.g. used in package descriptor) to make this libgpiod-development branch discernible from the mainstream packages

hangeza committed:
removed several debug outputs and decreased the status2 LED flash time to 10ms

hangeza committed:
improved preinst script to check whether muonuser exists and already is in gpio group

hangeza committed:
Merge branch 'dev' into migrate-daemon-to-libgpiod

hangeza committed:
applied clang-format

hangeza committed:
added constant with deadtime increment and increased this value to 50us

hangeza committed:
removed shadowed redeclaration

hangeza committed:
removed blank line

SimonGLM committed:
Merge commit '5da24ae77037c3344817bf227996218eee60b371' into v2.1.0

SimonGLM committed:
Merge pull request #90 from MuonPi/v2.1.0

V2.1.0

SimonGLM committed:
Updated changelog

SimonGLM committed:
Merge branch 'v2.1.0' - Updated changelogs

hangeza committed:
Merge branch 'dev' into migrate-daemon-to-libgpiod

hangeza committed:
applied clang-format

hangeza committed:
Merge tag 'v2.1.0' into migrate-daemon-to-libgpiod

hangeza committed:
improved event timestamp handling
